packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/FLXProgressListenerTests.kt

```diff
@@ -0,0 +1,303 @@
+            val flow = realm.syncSession.progressAsFlow(Direction.UPLOAD, ProgressMode.INDEFINITELY)
+            assertFailsWith<RuntimeException> {
+                flow.collect {
+                    @Suppress("TooGenericExceptionThrown")
+                    throw RuntimeException("Crashing progress flow")
+                }
+            }
+
+            withTimeout(TIMEOUT) {
+                flow.first { it.isTransferComplete }
+            }
+        }
+    }
+
+    @Test
+    fun worksAfterCancel() = runBlocking {
+        Realm.open(createSyncConfig(app.createUserAndLogIn())).use { writerRealm ->
+            Realm.open(createSyncConfig(app.createUserAndLogin())).use { realm ->
+                // Setup a flow that we are just going to cancel
+                val flow = realm.syncSession.progressAsFlow(Direction.DOWNLOAD, ProgressMode.INDEFINITELY)
+
+                supervisorScope {
+                    val mutex = Mutex(true)
+                    val task = async {
+                        flow.collect {
+                            mutex.unlock()
+                        }
+                    }
+                    // Await the flow actually being active, this requires actual data transfer as
+                    // we arent guaranteed any initial events.
+                    writerRealm.writeSampleData(TEST_SIZE, timeout = TIMEOUT)
+                    mutex.lock()
+                    task.cancel()
+                }
+
+                // Verify that progress listeners still work
+                withTimeout(TIMEOUT) {
+                    val task = async { flow.first { it.isTransferComplete } }
+                    // Trigger data transfer to ensure we get an event at some point
+                    writerRealm.writeSampleData(TEST_SIZE, timeout = TIMEOUT)
+                    task.await()
+                }
+            }
+        }
+    }
+
+    @Test
+    fun completesOnClose() = runBlocking {
+        val channel = TestChannel<Boolean>(capacity = 5, onBufferOverflow = BufferOverflow.DROP_OLDEST, failIfBufferIsEmptyOnCancel = false)
+        TestApp("completesOnClose", TEST_APP_FLEX).use { app ->
+            val user = app.createUserAndLogIn()
+            val realm = Realm.open(createSyncConfig(user))
+            try {
+                val flow = realm.syncSession.progressAsFlow(Direction.UPLOAD, ProgressMode.INDEFINITELY)
+                val job = async {
+                    withTimeout(10.seconds) {
+                        flow.collect {
+                            channel.trySend(true)
+                        }
+                    }
+                }
+                realm.writeSampleData(TEST_SIZE, timeout = TIMEOUT)
+                // Wait for Flow to start, so we do not close the Realm before
+                // `flow.collect()` can be called.
+                channel.receiveOrFail()
+                realm.close()
+                job.await()
```

Comment:
This could hang forever if the test fails in some unexpected way.

Reply:
I have created a ticket to address this after the PR. We should revisit these tests.